spec/codecs/auto_flush_spec.rb

```diff
@@ -4,109 +4,102 @@
+    Thread.abort_on_exception = true
+  end
+
+  after(:all) do
+    Thread.abort_on_exception = @thread_abort
+  end
+
   let(:flushable)  { Mlc::AutoFlushTracer.new }
-  let(:flush_wait) { 0.1 }
+  let(:flush_wait) { 0.25 }
 
   describe LogStash::Codecs::AutoFlush do
     subject { described_class.new(flushable, flush_wait) }
 
     context "when initialized" do
-      it "#pending? is false" do
-        expect(subject.pending?).to be_falsy
-      end
-
       it "#stopped? is false" do
-        expect(subject.stopped?).to be_falsy
-      end
-
-      it "#finished? is true" do
-        expect(subject.finished?).to be_truthy
+        expect(subject.stopped?).to be_falsey
       end
     end
 
-    context "when started" do
-      let(:flush_wait) { 20 }
-
+    context "when starting once" do
       before { subject.start }
       after  { subject.stop }
 
-      it "#pending? is true" do
-        expect(subject.pending?).to be_truthy
-      end
-
-      it "#stopped? is false" do
-        expect(subject.stopped?).to be_falsy
-      end
-
-      it "#finished? is false" do
-        expect(subject.finished?).to be_falsy
+      it "calls auto_flush on flushable" do
+        sleep flush_wait + 0.5
+        expect(flushable.trace_for(:auto_flush)).to be_truthy
       end
     end
 
-    context "when finished" do
-      before do
-        subject.start
-        sleep flush_wait + 0.1
-      end
-
+    context "when starting multiple times" do
+      before { subject.start }
       after  { subject.stop }
 
       it "calls auto_flush on flushable" do
-        expect(flushable.trace_for(:auto_flush)).to be_truthy
-      end
+        expect(flushable.trace_for(:auto_flush)).to be_falsey
+        sleep 0.1
+        subject.start
+        expect(flushable.trace_for(:auto_flush)).to be_falsey
+        sleep 0.1
+        subject.start
+        expect(flushable.trace_for(:auto_flush)).to be_falsey
 
-      it "#pending? is false" do
-        expect(subject.pending?).to be_falsy
+        sleep flush_wait + 0.5
+        expect(flushable.trace_for(:auto_flush)).to be_truthy
       end
+    end
 
-      it "#stopped? is false" do
-        expect(subject.stopped?).to be_falsy
+    context "when retriggering during execution" do
+      let(:flush_wait) { 1 }
+      let(:delay)      { 1 }
+      before do
+        flushable.set_delay(delay)
+        subject.start
       end
+      after  { subject.stop }
 
-      it "#finished? is true" do
-        expect(subject.finished?).to be_truthy
+      it "calls auto_flush twice on flushable" do
+        now = Time.now.to_f
+        combined_time = flush_wait + delay
+        expect(flushable.trace_for(:auto_flush)).to be_falsey
+        sleep flush_wait + 0.25 # <---- wait for execution to start
+        subject.start          # <---- because of the semaphore, this waits for delay = 2 seconds, then starts a new thread
+        sleep combined_time + 0.5   # <---- wait for more than 4 seconds = flush_wait + delay
```

Comment:
I guess so.  Sequencing is more designed to simulate other threads doing stuff while the RSpec thread is blocked.  I will try.

BTW, Jenkins took 5+ minutes to run all the spec but on my MPB it takes 30+ seconds wtf?